database/email.go

```diff
@@ -0,0 +1,200 @@
+	}
+)
+
+// EmailCreate creates an email message in the DB which is waiting to be sent.
+func (db *DB) EmailCreate(ctx context.Context, m EmailMessage) error {
+	fields, err := bson.Marshal(m)
+	if err != nil {
+		return err
+	}
+	_, err = db.staticEmails.InsertOne(ctx, fields)
+	if err != nil {
+		return errors.AddContext(err, "failed to Insert")
+	}
+	return nil
+}
+
+// EmailLockAndFetch locks up to batchSize records with the given lockId and
+// returns up to batchSize locked entries. Some of the returned entries might
+// not have been locked during the current execution.
+func (db *DB) EmailLockAndFetch(ctx context.Context, lockID string, batchSize int64) ([]EmailMessage, error) {
+	// Find out how many entries are already locked by this id. Maybe we don't
+	// need to lock any additional ones.
+	filter := bson.M{
+		"locked_by": lockID,
+		"failed":    bson.M{"$ne": true},
+		"sent_at":   nil,
+	}
+	count, err := db.staticEmails.CountDocuments(ctx, filter)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to count locked email messages")
+	}
+	// Lock some more entries in order to fill the batch.
+	if count < batchSize {
+		// As MongoDB doesn't have an "update up to N entries" operation, what
+		// we do here is fetch the ids of the desired number of entries to
+		// lock and then lock them by ids.
+		ids, err := db.fetchUnlockedMessageIDs(ctx, batchSize-count)
+		if err != nil {
+			return nil, errors.AddContext(err, "failed to fetch message ids to lock")
+		}
+		err = db.lockMessages(ctx, lockID, ids)
+		if err != nil {
+			return nil, errors.AddContext(err, "failed to lock messages")
+		}
+	}
+	// Fetch up to batchSize messages already locked with lockID.
+	opts := options.Find()
+	opts.SetLimit(batchSize)
+	c, err := db.staticEmails.Find(ctx, filter, opts)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to fetch emails")
+	}
+	defer func() {
+		if errDef := c.Close(ctx); errDef != nil {
+			db.staticLogger.Traceln("Error on closing DB cursor.", errDef)
+		}
+	}()
+	var msgs []EmailMessage
+	for c.Next(ctx) {
+		var m EmailMessage
+		if err = c.Decode(&m); err != nil {
+			return nil, errors.AddContext(err, "failed to parse value from DB")
+		}
+		msgs = append(msgs, m)
+	}
+	return msgs, nil
+}
+
+// fetchUnlockedMessageIDs is a helper method that fetches the ids of up to num
+// unlocked email messages waiting to be sent.
+func (db *DB) fetchUnlockedMessageIDs(ctx context.Context, num int64) ([]primitive.ObjectID, error) {
+	filter := bson.M{
+		"locked_by": "",
+		"failed":    bson.M{"$ne": true},
+		"sent_at":   nil,
+	}
+	opts := options.Find()
+	opts.SetLimit(num)
+	c, err := db.staticEmails.Find(ctx, filter, opts)
+	if err != nil {
+		return nil, errors.AddContext(err, "failed to fetch ids")
+	}
+	defer func() {
+		if errDef := c.Close(ctx); errDef != nil {
+			db.staticLogger.Traceln("Error on closing DB cursor.", errDef)
+		}
+	}()
+	var ids []primitive.ObjectID
+	for c.Next(ctx) {
```

Comment:
This could be refactored to a method that returns both the ids and the messages.

Reply:
The reason for the current approach is that I don't have a way to lock a limited number of records, so I want to lock them by their ids. Which means that another server doing the same thing at the same time might decide to update the same records. If that happens and `fetchUnlockedMessageIDs` returns the records instead of the ids, those emails will get sent twice. With the current setup the messages updated by both servers will be sent just once and one of the servers will send fewer emails in its current batch, which is not an issue.

There might be something I'm missing here. It will be easier to discuss that on Discord.

Reply:
You have the same for loop that iterates over `for c.Next(ctx)` twice https://github.com/SkynetLabs/skynet-accounts/pull/72/files/7f099e6c21a8226a5ccc9f0155e67b09d5b46e06#diff-23fbde68a3027d369c3bf5fcd43e6d5cce2830e5108303066a7d9fc14e0c1358R90

Both for loops decode the EmailMessage. One loop creates a slice of EmailMessages, and one loop creates a slice of ObjectIDs. 

You could just have one helper method that returns both. The looping of the `c.Next(ctx)` feels like something that might happen for a number of reasons so might be worth having pulled out. 

Not blocking, just a thought.

Reply:
I think I misunderstood the original idea. Moving this to a helper makes sense. Will add it in the next PR.